### PR TITLE
TypeScript type definitions

### PR DESCRIPTION
**Pull request by alexisfontaine**
Connects to #94.

- [x] `Time`
- [x] `decide()`
- [x] `createClient()`
    - [x] `.cfg`
    - [x] `.addAgentContextOperations()`
    - [x] `.createAgent()`
    - [x] `.deleteAgent()`
    - [x] `.getAgent()`
    - [x] `.getAgentContext()`
    - [x] `.getAgentContextOperations()`
    - [x] `.getAgentDecisionTree()`
    - [x] `.listAgents()`
    - [x] `.destroyAgent()`
    - [x] `.computeAgentDecision()`
    - [x] `.getAgentInspectorUrl()`
    - [x] `.getSharedAgentInspectorUrl()`
    - [x] `.deleteSharedAgentInspectorUrl()`

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -13,6 +13,7 @@
     "url": "https://github.com/craft-ai/craft-ai-client-js/issues"
   },
   "main": "lib/index.js",
+  "typings": "src/index.d.ts",
```
Comment (cloderic) [inline package.json]: You should add some minimal checks on the file using `tsc`

cf. https://github.com/craft-ai/most-utils/blob/83e14ac52bbadd0282fa4b83e2aa9938a3cb4b94/packages/most-buffer/package.json#L34-L41

File: src/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+  lastTimestamp: number
+  creationDate: number
+  lastTreeUpdate: number
+  lastContextUpdate: number
+}
+
+export interface DecisionTree<F extends Features> extends Versionable {
+  configuration: Configuration<F>
+  timestamp?: number
+  trees?: {
```
Comment (cloderic) [inline src/index.d.ts]: In which case `trees` is not defined ?

File: src/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+  id: string
+  configuration: Configuration<F>
+  firstTimestamp: number
+  lastTimestamp: number
+  creationDate: number
+  lastTreeUpdate: number
+  lastContextUpdate: number
+}
+
+export interface DecisionTree<F extends Features> extends Versionable {
```
Comment (cloderic) [inline src/index.d.ts]: We hav defined JSON Schema for these, maybe there's a way to convert the json schema to the `.d.ts`

File: src/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+
+export interface Decision<F extends Features> extends Versionable {
+  context: Context<F>
+  output: {
+    [feature: string]: {
+      predicted_value?: string | number
+      confidence?: number
+      standard_deviation?: number
+      decision_rules: DecisionRule.Any[]
+      error?: {
```
Comment (cloderic) [inline src/index.d.ts]: Hum this is thrown no returned

File: src/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+   * @param agentId Id of the agent
+   */
+  deleteAgent<F extends Features> (agentId: string): Promise<Agent<F>>
+
+  /**
+   * Deletes the public inspector URL for the given agent id.
+   *
+   * @param agentId Id of the agent
+   */
+  deleteSharedAgentInspectorUrl (agentId: string): Promise<true>
```
Comment (cloderic) [inline src/index.d.ts]: Promise<undefined> ?

File: src/index.d.ts [hunk 6]
```diff
@@ -0,0 +1,280 @@
+interface Versionable {
+  _version?: string
+}
+
+export namespace Feature {
```
Comment (cloderic) [inline src/index.d.ts]: In craft ai longue this is called `Property`

File: src/index.d.ts [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+ * @param tree Decision tree retrieved from craft ai (see `client.getAgentDecisionTree()`)
+ * @param timeOrContextOperation A list of partial context operation or an instance of `Time`
+ */
+export function decide<F extends Features> (tree: DecisionTree<F>, ...timeOrContextOperation: (Partial<Context<F>> | Time)[]): Decision<F>
+
+/**
+ * Creates a client instance of craft ai.
+ *
+ * A client is bound to a project.
+ * The configuration must include at least the write access token to a craft ai project.
```
Comment (cloderic) [inline src/index.d.ts]: you can create a client with only read access

File: src/index.d.ts (hunk 3 above)
Comment (alexisfontaine) [inline src/index.d.ts]: I used https://github.com/bcherny/json-schema-to-typescript as a starting point. Beyond some mistakes in the schema, the relationships between the types/interfaces were too basic for a proper type-checking.

File: src/index.d.ts (hunk 4 above)
Comment (alexisfontaine) [inline src/index.d.ts]: You're right, `decide` is throwing whenever it has the property `error`. So it never returns a decision with it.

File: src/index.d.ts (hunk 2 above)
Comment (alexisfontaine) [inline src/index.d.ts]: I was relying on the JSON Schema. I checked and it's either a `DecisionTreeLeaf` or a `DecisionTreeNode` with no property `decision_rule`.

File: src/index.d.ts (hunk 5 above)
Comment (alexisfontaine) [inline src/index.d.ts]: I changed its return value and updated the type definition.

**Pull request merged**